AP_Scripting: add support for binding dependency, add plane specific in_vtol_mode binding

IamPete1 opened the pull request:
This adds as the ability to conditionally compile bindings. This paves the way for many more vehicle specific functions (we might even want to move a some of the current vehicle ones to copter only). We should probably go back and add the appropriate depends to the various existing bindings.

Hopefully this also helps get scripting on smaller boards as we can now do scripting without having to have terrain available.

Required for https://github.com/ArduPilot/ardupilot/pull/15708 and/or https://github.com/ArduPilot/ardupilot/pull/13693

This also adds `in_vtol_mode()` binding because I need it and it makes a nice example of how the dependency is setup in the description

IamPete1 commented:
Tested and fixed this. The generator is working fine. However `APM_BUILD_TYPE(APM_BUILD_ArduPlane)` is always `APM_BUILD_UNKNOWN` so the new quadplane bindings are never included no matter the vehicle.

IamPete1 commented:
also getting the linux error: `build/linux/libraries/AP_Scripting/lua_generated_bindings.h: library header uses vehicle-dependent macros`

tridge commented:
I like the depends keyword!

IamPete1 commented:
This should now works as expected in SITL. I have added all files in `/build` to be vehicle dependent. At the moment that is only the scripting generated bindings. We are including `../ArduPlane/quadplane.h` in `lua_generated_bindings.h` on plane. This is only included in the following files: 

- `lua_generated_bindings.cpp`
- `lua_bindings.cpp`
- `lua_repl.cpp`
- `lua_scripts.cpp`

Hopefully that is not too painfull. The other solution to get to quad-plane would be to alter the binding gen quite abit more so we can go via plane.

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -273,12 +274,29 @@ void handle_header(void) {
+  char * key_word = next_token();
+  char *depends = NULL;
+  if (key_word != NULL) {
+    if (strcmp(key_word, keyword_depends) != 0) {
+      error(ERROR_HEADER, "Headers only support depends keyword, got %s", key_word);
+    }
+    depends = strtok(NULL, "");
+    if (depends == NULL) {
+      error(ERROR_HEADER, "Expected a depends string for %s", name);
+    }
```
```suggestion
    if (strcmp(key_word, keyword_depends) == 0) {
        depends = strtok(NULL, "");
        if (depends == NULL) {
          error(ERROR_HEADER, "Expected a depends string for %s", name);
        }
    } else {
      error(ERROR_HEADER, "Recieved an unsupported keyword on a header: %s", key_word);
    }
```

We should be using this pattern instead of handling everything first, rather then checking for the one case you currently care about. The problem is when we come back to add something else to this the code churn will be much larger. We should strive to keep a general pattern of handling things first, then falling to the error state last.

IamPete1 replied:
made suggested change

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -273,12 +274,29 @@ void handle_header(void) {
   node->line = state.line_num;
   node->name = (char *)allocate(strlen(name) + 1);
   strcpy(node->name, name);
+  if (depends != NULL) {
```
This should be handled right in where the token is found above, rather then just being a character pointer that gets handled later.

IamPete1 replied:
done

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -273,12 +274,29 @@ void handle_header(void) {
   node->name = (char *)allocate(strlen(name) + 1);
   strcpy(node->name, name);
+  if (depends != NULL) {
+    node->dependency = (char *)allocate(strlen(depends) + 1);
+    strcpy(node->dependency, depends);
```
```suggestion
    string_copy(&(node->dependency), depends);
```

We have a helper to wrap this up and be clearer.

IamPete1 replied:
fixed, I had to move `string_copy` up in the file to get it to work

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -970,53 +957,84 @@ void emit_userdata_allocators(void) {
-    fprintf(source, "}\n\n");
+    fprintf(source, "}\n");
+    if (node->dependency != NULL) {
+      fprintf(source, "#endif\n");
```
```suggestion
      fprintf(source, "#endif // %s\n", node->dependency);
```
It's helpful sometimes (particularly when debugging) to have the comment indicate what we just finished checking

WickedShell replied:
Actually emitting this header/footer for dependencies is so common maybe we should have a helper for `start_dependency()` and `end_dependency()`? It's readable as is, I'm just noting that there seems to be a lot of repeated bits at the moment.

IamPete1 replied:
moved to a `start_dependency()` and `end_dependency()` helpers

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -970,53 +957,84 @@ void emit_userdata_allocators(void) {
-    fprintf(source, "}\n\n");
+    fprintf(source, "}\n");
+    if (node->dependency != NULL) {
+      fprintf(source, "#endif\n");
```
Same as above.

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -970,53 +957,84 @@ void emit_userdata_allocators(void) {
-    fprintf(source, "}\n\n");
+    fprintf(source, "}\n");
+    if (node->dependency != NULL) {
+      fprintf(source, "#endif\n");
```
same as above

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -970,53 +957,84 @@ void emit_userdata_allocators(void) {
-    fprintf(source, "}\n\n");
+    fprintf(source, "}\n");
+    if (node->dependency != NULL) {
+      fprintf(source, "#endif\n");
```
same as above

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -970,53 +957,84 @@ void emit_userdata_allocators(void) {
     fprintf(header, "int new_%s(lua_State *L);\n", node->sanatized_name);
     fprintf(header, "%s * check_%s(lua_State *L, int arg);\n", node->name, node->sanatized_name);
+    if (node->dependency != NULL) {
+      fprintf(source, "#endif\n");
```
same as above

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -1316,10 +1334,16 @@ void emit_userdata_fields() {
       field = field->next;
     }
+    if (node->dependency != NULL) {
+      fprintf(source, "#endif\n");
```
same as above

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -1636,7 +1664,12 @@ void emit_userdata_method(const struct userdata *data, const struct method *meth
-  fprintf(source, "}\n\n");
+  fprintf(source, "}\n");
+  if (data->dependency != NULL) {
+    fprintf(source, "#endif\n");
```
same as above

WickedShell commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -1745,15 +1781,21 @@ void emit_userdata_metatables(void) {
-
+    fprintf(source, "};\n");
+    if (node->dependency != NULL) {
+      fprintf(source, "#endif\n");
```
same as above

WickedShell commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -148,8 +148,7 @@ singleton RangeFinder method get_pos_offset_orient Vector3f Rotation'enum ROTATI
 
-depends AP_TERRAIN_AVAILABLE 1 Scripting requires terrain to be available
-
+singleton AP_Terrain depends defined(AP_TERRAIN_AVAILABLE) && AP_TERRAIN_AVAILABLE == 1
```
The downside to this is that previously scripting implied that terrain was available, where as now your script has to cope with the idea that terrain may not be there. Do we need a form that is a hard requirement, as well as an optional feature, or do we want to embrace allowing terrain to just be removed from scripting builds?

IamPete1 replied:
added a exist method so we can now cope with singletons not being there more gracefully.

WickedShell reviewed (changes requested):
Not a complete review, ran out of time before the call, I definitely want to look at the generated output for this one, and test some edge cases.

IamPete1 commented:
check that total number of objects does not increase more than 3 or 4, check files are not rebuilt on a second build such as `./waf plane` then  `./waf copter`

IamPete1 commented:
Fixed review issues, tested build and as far as I can tell the changes to `ap_library.py` are not causing un-needed file builds.

We should probably add some function that you can ask if the singleton your trying to use exists. Apparently a `pcall` is not enough to allow one to check without error. Maybe `has(string)`? doing something like `singleton:exist()` will still need all the meta tables and such to be setup. I guess we could go the other way and to `exist:singleton()`

IamPete1 commented:
This now also adds a exist singleton with a method for each other singleton, this allows the script to check if a singletons is available. For example:

```
function update()

  if exist:quadplane() then
    gcs:send_text(0, tostring(quadplane:in_vtol_mode()))
  else 
    gcs:send_text(0, "QuadPlane not available")
  end

  return update, 1000
end

return update()
```

This will allow scripts to better cope with the possibility that not everything will be available, and give useful errors. The methods are auto generated for any new singletons.

Currently there is no way to handle this without hitting a `x not supported on this firmware` hard error even in a `pcall()`

IamPete1 commented:
This is the diff of the generated bindings, Master vs this. [generated.txt](https://github.com/ArduPilot/ardupilot/files/5673820/generated.txt)

tridge commented:
@WickedShell I like this idea, just needs you to look at it

IamPete1 commented:
Just double checked `./waf build all` only a small in increase in total build items, +5

IamPete1 commented:
rebased

IamPete1 commented:
dropped exist method because `exist:quadplane()` is equivalent to `quadplane ~= null`. The exist method would work for stuff that comes up at run time, or if a singleton goes away that the existing stuff cannot, but that should not be happening anyway.

Also rebased
